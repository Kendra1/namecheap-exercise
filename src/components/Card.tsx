import { IMovie } from "@/types/movies";
import React, { Dispatch, FunctionComponent, SetStateAction } from "react";

export const Card: FunctionComponent<{
    movie: IMovie;
    setSelectedMovie: Dispatch<SetStateAction<IMovie | null>>;
}> = ({ movie, setSelectedMovie }) => {
    const movieYear: number | string =
        new Date(movie.releaseDate).getFullYear() || "N/A";
    return (
        <div
            className="card movie-card"
            onClick={() => {
                setSelectedMovie(movie);
            }}
        >
            <img className="card-img-top poster" src={movie.posterPath} />
            <div className="card-body text-center">
                <div className="card-title title font-weight-bold fs-3 fw-bold">
                    {movie.title}
                </div>
                <p className="fs-5 movie-year">{movieYear}</p>
                <p className="fs-5 movie-genres">{movie.genres.join("/")}</p>
                <div className="card-text overview">{movie.overview}</div>
            </div>
        </div>
    );
};
