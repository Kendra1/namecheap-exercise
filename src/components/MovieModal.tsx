import { IMovie } from "@/types/movies";
import React, { FunctionComponent } from "react";
import { Modal, CloseButton, Container } from "react-bootstrap";

export const MovieModal: FunctionComponent<{
    movie: IMovie | null;
    closeModal: () => void;
}> = ({ movie, closeModal }) => {
    const movieYear: number | string = movie
        ? new Date(movie.releaseDate).getFullYear()
        : "N/A";
    return (
        movie && (
            <Modal
                show={!!movie}
                size="lg"
                onHide={closeModal}
                aria-labelledby="contained-modal-title-vcenter"
                centered
            >
                <div className="button-container">
                    <CloseButton onClick={closeModal} />
                </div>
                <Modal.Header className="header pt-0">
                    <div className="fs-1 text-center w-100">{movie.title}</div>
                </Modal.Header>
                <Modal.Body className="content">
                    <img className="poster" src={movie.backdropPath} />
                    <Container className="info text-center">
                        <h5>
                            {movieYear}, {movie.genres.join("/")},
                            {` ${movie.direction}`}
                        </h5>
                        <p>{movie.overview}</p>
                    </Container>
                </Modal.Body>
            </Modal>
        )
    );
};
