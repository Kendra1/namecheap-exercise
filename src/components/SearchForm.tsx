import { FunctionComponent, useState, KeyboardEvent } from "react";
import { Button, Col, Form } from "react-bootstrap";

export const SearchForm: FunctionComponent<{
    searchMovies: (searchQuery: string) => Promise<void>;
}> = ({ searchMovies }) => {
    const [searchQuery, setSearchQuery] = useState<string>("");

    const handleSearchMovies = () => {
        searchMovies(searchQuery);
    };

    const handleSearch = (event: KeyboardEvent<HTMLElement>) => {
        if (event.key === "Enter") {
            event.preventDefault();
            handleSearchMovies();
        }
    };

    return (
        <Form className="row mb-3">
            <div className="mb-3 input-group w-50 col-6 offset-3 ">
                <span className="input-group-append">
                    <i className="fa fa-search">
                        <svg
                            xmlns="http://www.w3.org/2000/svg"
                            width="16"
                            height="16"
                            fill="currentColor"
                            className="bi bi-search"
                            viewBox="0 0 16 16"
                        >
                            <path d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z" />
                        </svg>
                    </i>
                </span>
                <Form.Control
                    onKeyDown={handleSearch}
                    className="search-field"
                    value={searchQuery}
                    onChange={(e) => setSearchQuery(e.target.value)}
                />
            </div>
            <Col xs={6} className="offset-3 text-center">
                <Button className="btn-dark" onClick={handleSearchMovies}>
                    Search
                </Button>
            </Col>
        </Form>
    );
};
