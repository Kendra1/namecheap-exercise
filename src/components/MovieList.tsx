import { IMovie } from "@/types/movies";
import React, { Dispatch, FunctionComponent, SetStateAction } from "react";
import { Container, Row, Col } from "react-bootstrap";
import { Card } from "./Card";

export const MovieList: FunctionComponent<{
    movies: IMovie[];
    setSelectedMovie: Dispatch<SetStateAction<IMovie | null>>;
}> = ({ movies, setSelectedMovie }) => {
    return (
        <Container>
            <Row>
                {movies.map((movie: IMovie) => (
                    <Col key={movie.id} xs={12} sm={6} lg={4}>
                        <Card
                            movie={movie}
                            setSelectedMovie={setSelectedMovie}
                        />
                    </Col>
                ))}
            </Row>
        </Container>
    );
};
