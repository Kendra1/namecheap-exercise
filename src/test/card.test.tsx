import "@testing-library/jest-dom";
import renderer, {
    ReactTestRenderer,
    ReactTestRendererJSON,
} from "react-test-renderer";
import { Card } from "../components/Card";

const movie = {
    adult: false,
    backdropPath:
        "https://image.tmdb.org/t/p/w500/22z44LPkMyf5nyyXvv8qQLsbom.jpg",
    genres: ["Horror", "Mystery", "Thriller"],
    id: 631842,
    originalLanguage: "en",
    originalTitle: "Knock at the Cabin",
    direction: "M. Night Shyamalan",
    overview:
        "While vacationing at a remote cabin, a young girl and her two fathers are taken hostage by four armed strangers who demand that the family make an unthinkable choice to avert the apocalypse. With limited access to the outside world, the family must decide what they believe before all is lost.",
    popularity: 4969.06,
    posterPath:
        "https://image.tmdb.org/t/p/w500/dm06L9pxDOL9jNSK4Cb6y139rrG.jpg",
    releaseDate: new Date("2023-02-01"),
    title: "Knock at the Cabin",
    video: false,
    voteAverage: 6.6,
    voteCount: 778,
};

let component: ReactTestRenderer;
let tree: ReactTestRendererJSON;
let root: any;
const baseProps = {
    movie: movie,
    setSelectedMovie: jest.fn(),
};

describe("Card succesfull", () => {
    beforeEach(() => {
        component = renderer.create(<Card {...baseProps}></Card>);
        tree = component.toJSON() as ReactTestRendererJSON;
        root = component.root;
    });
    it("should render the component ", () => {
        expect(tree).toMatchSnapshot();
    });
    it("should call set selected movie when clicked on", () => {
        const cardDiv = root.findByProps({
            className: "card movie-card",
        });

        cardDiv.props.onClick();
        expect(baseProps.setSelectedMovie).toHaveBeenCalled();
    });
    it("should contain a proper source for displayed image", () => {
        const img = root.findByProps({
            className: "card-img-top poster",
        });

        expect(img.props.src).toBe(movie.posterPath);
    });
    it("should display proper movie title", () => {
        const titleDiv = root.findByProps({
            className: "card-title title font-weight-bold fs-3 fw-bold",
        });

        expect(titleDiv.props.children).toBe(movie.title);
    });
    it("should display proper year", () => {
        const yearParagraph = root.findByProps({
            className: "fs-5 movie-year",
        });

        expect(yearParagraph.props.children).toBe(
            movie.releaseDate.getFullYear()
        );
    });
    it("should display proper genres", () => {
        const genresParagraph = root.findByProps({
            className: "fs-5 movie-genres",
        });

        expect(genresParagraph.props.children).toBe(movie.genres.join("/"));
    });
    it("should display proper overview", () => {
        const overviewDiv = root.findByProps({
            className: "card-text overview",
        });

        expect(overviewDiv.props.children).toBe(movie.overview);
    });
});
