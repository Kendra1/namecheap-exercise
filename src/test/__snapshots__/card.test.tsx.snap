// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card succesfull should render the component  1`] = `
<div
  className="card movie-card"
  onClick={[Function]}
>
  <img
    className="card-img-top poster"
    src="https://image.tmdb.org/t/p/w500/dm06L9pxDOL9jNSK4Cb6y139rrG.jpg"
  />
  <div
    className="card-body text-center"
  >
    <div
      className="card-title title font-weight-bold fs-3 fw-bold"
    >
      Knock at the Cabin
    </div>
    <p
      className="fs-5 movie-year"
    >
      2023
    </p>
    <p
      className="fs-5 movie-genres"
    >
      Horror/Mystery/Thriller
    </p>
    <div
      className="card-text overview"
    >
      While vacationing at a remote cabin, a young girl and her two fathers are taken hostage by four armed strangers who demand that the family make an unthinkable choice to avert the apocalypse. With limited access to the outside world, the family must decide what they believe before all is lost.
    </div>
  </div>
</div>
`;
